_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,192 @@
+public class DynamicKeyFunction
+    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {
+
+  /* Simplified */
+  List<Rule> rules = /* Rules that are initialized somehow.*/;
+  ...
+}
+```
+
+Obviously, adding rules to this list is possible directly inside the code of the Flink Job at the stage of its initialization (create a List object, use it's `add` method). A major drawback of doing so is that it will require recompilation of the job with each rule modification. In the real Fraud Detection system, you can expect to modify these rules on a frequent basis, which makes this approach unacceptable from the point of view of business and operational requirements - a different approach is needed.
+
+Let's next have a look at the example of DSL of one of the sample rules presented in the previous post:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rule-dsl.png" width="800px" alt="Figure 1: Rules DSL"/>
+<br/>
+<i><small>Figure 1: Rules DSL</small></i>
+</center>
+<br/>
+![](assets/rule-dsl.png)
+
+Earlier we have already seen how part of this definition, namely `groupingKeyNames`, is used in `DynamicKeyFunction` to extract message keys. At the same time, all other parameters of this rule are necessary within `DynamicAlertFunction` - they define the actual logic of performed operations and it's parameters (such as the alert triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use a special data distribution mechanism in Apache Flink - [events broadcast](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/broadcast_state.html). Let’s turn to the Job Graph of the system that we are building:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph"/>
+<br/>
+<i><small>Figure 2: Job Graph</small></i>
+</center>
+<br/>
+![](assets/job-graph.png)
+
+On it you can see the main blocks of our Transactions processing pipeline:<br>
+
+* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>
+    * The forward operator following it means that all data consumed by one of the parallel instances of Transaction Source operator will be transferred to exactly one instance of the subsequent DynamicKeyFunction operator, without redistribution. It also indicates the same level of parallelism of two connected operators (12 in this case).
+
+ <center>
+ <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across operators : FORWARD"/>
+ <br/>
+ <i><small>Figure 3: Message passing across operators : FORWARD</small></i>
+ </center>
+ <br/>
+![](assets/forward.png)
+
+* **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator performs `hashing` of this dynamic key and corresponding data sharding
```

Comment:
Hmm... https://www.grammarly.com/blog/comma-before-which/.

Reply:
Hard to say what the "sentence" is here and in enumerations of that style with sub-bullets. I guess, it would even work without the "which", but it all depends how you write this block in general. I won't fight over a comma ;)

Reply:
My understanding is that a "task" is already one of the parallel instances and it cannot be further broken down, so "among all parallel instances of the following task" is not an applicable description.

Reply:
Regarding the comma, if you add one before "which", you may also want to add one before "that" which is essentially the same.